src/connector/src/sink/mqtt.rs

```diff
@@ -244,24 +333,251 @@ impl Drop for MqttSinkWriter {
+                &self.topic_index_path,
+                self.topic.as_deref(),
+                &row,
+            ) {
+                Some(s) => s,
+                None => {
+                    tracing::error!("topic field not found in row, skipping: {:?}", row);
+                    return Ok(());
+                }
+            };
 
-    async fn write_one(&mut self, _k: Option<Self::K>, v: Option<Self::V>) -> Result<()> {
-        match v {
-            Some(v) => self
-                .client
-                .publish(&self.config.common.topic, self.qos, self.retain, v)
+            let v = encoder.encode(row)?;
+
+            self.client
+                .publish(topic, self.qos, self.retain, v)
                 .await
                 .context("mqtt sink error")
-                .map_err(SinkError::Mqtt),
-            None => Ok(()),
+                .map_err(SinkError::Mqtt)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn get_topic_from_index_path<'s>(
+    path: &[usize],
+    default_topic: Option<&'s str>,
+    row: &'s RowRef<'s>,
+) -> Option<&'s str> {
+    if let Some(topic) = default_topic
+        && path.is_empty()
+    {
+        Some(topic)
+    } else {
+        let mut iter = path.iter();
+        let scalar = iter
+            .next()
+            .and_then(|pos| row.datum_at(*pos))
+            .and_then(|d| {
+                iter.try_fold(d, |d, pos| match d {
+                    ScalarRefImpl::Struct(struct_ref) => {
+                        struct_ref.iter_fields_ref().nth(*pos).flatten()
+                    }
+                    _ => None,
+                })
+            });
+        match scalar {
+            Some(ScalarRefImpl::Utf8(s)) => Some(s),
+            _ => {
+                if let Some(topic) = default_topic {
+                    Some(topic)
+                } else {
+                    None
+                }
+            }
         }
     }
 }
+
+fn get_topic_field_index_path(schema: &Schema, topic_field: &str) -> Result<Vec<usize>> {
+    let mut iter = topic_field.split('.');
+    let mut path = vec![];
+    let dt =
+        iter.next()
+            .and_then(|field| {
+                schema
+                    .fields()
+                    .iter()
+                    .enumerate()
+                    .find(|(_, f)| f.name == field)
+                    .map(|(pos, f)| {
+                        path.push(pos);
+                        &f.data_type
+                    })
+            })
+            .and_then(|dt| {
+                iter.try_fold(dt, |dt, field| match dt {
+                    DataType::Struct(st) => {
+                        st.iter().enumerate().find(|(_, (s, _))| *s == field).map(
+                            |(pos, (_, dt))| {
+                                path.push(pos);
+                                dt
+                            },
+                        )
+                    }
+                    _ => None,
+                })
+            });
```

Comment:
Can you write some comments and unit test? This part seems hard to understand.

Reply:
@tabVersion The unit tests are here https://github.com/risingwavelabs/risingwave/blob/d88033ffd47cb111f41cb2785ec7f3d15bfb1dda/src/connector/src/sink/mqtt.rs#L462-L582 do you think any more are needed?